### PR TITLE
check if resolveType returns a promise

### PR DESCRIPTION
**Pull request by benawad**
This is blocking me from using `graphql-jit` atm. It doesn't expect resolveType to return a promise.

### CODE REVIEW

**Comment by MichalLytek** [comment]
>  It doesn't expect resolveType to return a promise.

1) This is against GraphQL JS runtime behavior:

https://github.com/graphql/graphql-js/tree/b4bff0ba9c15c9d7245dd68556e754c41f263289/tstypes/type/definition.d.ts#L419-L430

2) This doesn't solve the issue, only makes a workaround that it will work until you for some reason return a promise 😕 

It's better to raise an issue to fix in GraphQL-JIT rather that restricting even on `graphql-js` side.

**Comment by benawad** [comment]
> This is against GraphQL JS runtime behavior:

yeah I brought it to their attention.

> This doesn't solve the issue, only makes a workaround that it will work until you for some reason return a promise 😕

Either way, forcing every resolveType resolver to return a promise is worse for performance. Why not let the developer choose whether they want it to return a promise or not?

File: src/schema/schema-generator.ts [hunk 1]
```diff
@@ -663,12 +670,26 @@ export abstract class SchemaGenerator {
...
     resolveType: TypeResolver<TSource, TContext>,
   ): GraphQLTypeResolver<TSource, TContext> {
-    return async (...args) => {
-      const resolvedType = await resolveType(...args);
-      if (typeof resolvedType === "string") {
-        return resolvedType;
+    return (...args) => {
+      const resolvedTypeOrPromise = resolveType(...args);
+
+      if (isPromiseLike(resolveType)) {
```
Comment (MichalLytek) [inline src/schema/schema-generator.ts]: I think it should be `resolvedTypeOrPromise` - we are checking the return value of `resolveType` fn

File: src/schema/schema-generator.ts [hunk 2]
```diff
@@ -41,7 +41,13 @@ import {
   ConflictingDefaultValuesError,
   InterfaceResolveTypeError,
 } from "../errors";
-import { ResolverFilterData, ResolverTopicData, TypeResolver } from "../interfaces";
+import {
+  ResolverFilterData,
+  ResolverTopicData,
+  TypeResolver,
+  Maybe,
+  ClassType as ClassTypeType,
```
Comment (MichalLytek) [inline src/schema/schema-generator.ts]: Why do you need such alias?

**Comment by MichalLytek** [comment]
> Why not let the developer choose whether they want it to return a promise or not?

As you can see, introducing sync paths results in duplicated code. TypeGraphQL was not designed with this kind of optimization in mind but the vNext is:

https://github.com/MichalLytek/type-graphql/blob/16d482d9537b87f16c95198c52f929f71ae065a9/packages/core/src/runtime/RuntimeGenerator.ts#L25-L46

Sure we can merge this PR but don't expect any refactor or future async-related optimization as I don't have enough free time to actively support two versions of TypeGraphQL.

**Comment by benawad** [comment]
I'm cool waiting for vNext for this change. I just wasn't sure if this was going to be fixed with it.

**Comment by MichalLytek** [comment]
Every "maybe async" execution path will be optimized in vNext for sync cases with `completeValue` helper - I should write a rule "no await/.then()" for the `runtime` folder 😄

**Comment by PabloSzx** [comment]
what happened with this issue @MichalLytek ? right now it's impossible to use unions with TypeGraphQL & Mercurius due to this limitation

**Comment by MichalLytek** [comment]
@PabloSzx it's out of scope - please ask `graphql-jit` team to support and conform to GraphQL JS specification

**Comment by StarpTech** [comment]
GraphQL JS specification says either promise or value. The current direction here is to enforce promises.

**Pull request closed without merging**